2019-05-04 Dzianis Kahanovich <mahatma@eu.by>

PR target/40171
* config/i386/driver-i386.c (host_detect_local_cpu): Add "astune" mode (assembler mtune).
* config/i386/i386.h (ASM_SPEC_MTUNE): Detect & pass -mtune=native to the GNU assembler.
* config/i386/linux64.h (ASM_SPEC): Append ASM_SPEC_MTUNE.
* config/i386/linux.h (ASM_SPEC): Likewise.
* config/i386/mingw-w64.h (ASM_SPEC): Likewise.
* config/i386/sol2-10.h (ASM_SPEC): Likewise.
* config/i386/x86-64.h (ASM_SPEC): Likewise.

--- a/gcc/config/i386/driver-i386.c	2018-12-25 03:16:35.000000000 +0300
+++ b/gcc/config/i386/driver-i386.c	2018-12-25 15:33:20.638569343 +0300
@@ -429,7 +429,7 @@ const char *host_detect_local_cpu (int a
 
   unsigned int has_ptwrite = 0;
 
-  bool arch;
+  bool arch, tune, astune;
 
   unsigned int l2sizekb = 0;
 
@@ -437,8 +437,10 @@ const char *host_detect_local_cpu (int a
     return NULL;
 
   arch = !strcmp (argv[0], "arch");
+  tune = !arch && !strcmp (argv[0], "tune");
+  astune = !arch && !tune && !strcmp (argv[0], "astune");
 
-  if (!arch && strcmp (argv[0], "tune"))
+  if (!arch && !tune && !astune)
     return NULL;
 
   max_level = __get_cpuid_max (0, &vendor);
@@ -629,7 +631,7 @@ const char *host_detect_local_cpu (int a
       has_avx512vl = 0;
     }
 
-  if (!arch)
+  if (tune)
     {
       if (vendor == signature_AMD_ebx
 	  || vendor == signature_CENTAUR_ebx
@@ -760,7 +762,7 @@ const char *host_detect_local_cpu (int a
 	case 0x1c:
 	case 0x26:
 	  /* Bonnell.  */
-	  cpu = "bonnell";
+	  cpu = astune ? "core2" : "bonnell";
 	  break;
 	case 0x37:
 	case 0x4a:
@@ -768,7 +770,7 @@ const char *host_detect_local_cpu (int a
 	case 0x5a:
 	case 0x5d:
 	  /* Silvermont.  */
-	  cpu = "silvermont";
+	  cpu = astune ? "core2" : "silvermont";
 	  break;
 	case 0x5c:
 	case 0x5f:
@@ -791,37 +793,37 @@ const char *host_detect_local_cpu (int a
 	case 0x1f:
 	case 0x2e:
 	  /* Nehalem.  */
-	  cpu = "nehalem";
+	  cpu = astune ? "corei7" : "nehalem";
 	  break;
 	case 0x25:
 	case 0x2c:
 	case 0x2f:
 	  /* Westmere.  */
-	  cpu = "westmere";
+	  cpu =  astune ? "corei7" : "westmere";
 	  break;
 	case 0x2a:
 	case 0x2d:
 	  /* Sandy Bridge.  */
-	  cpu = "sandybridge";
+	  cpu =  astune ? "corei7" : "sandybridge";
 	  break;
 	case 0x3a:
 	case 0x3e:
 	  /* Ivy Bridge.  */
-	  cpu = "ivybridge";
+	  cpu = astune ? "corei7" : "ivybridge";
 	  break;
 	case 0x3c:
 	case 0x3f:
 	case 0x45:
 	case 0x46:
 	  /* Haswell.  */
-	  cpu = "haswell";
+	  cpu = astune ? "corei7" : "haswell";
 	  break;
 	case 0x3d:
 	case 0x47:
 	case 0x4f:
 	case 0x56:
 	  /* Broadwell.  */
-	  cpu = "broadwell";
+	  cpu = astune ? "corei7" :"broadwell";
 	  break;
 	case 0x4e:
 	case 0x5e:
@@ -829,10 +831,12 @@ const char *host_detect_local_cpu (int a
 	case 0x8e:
 	case 0x9e:
 	  /* Kaby Lake.  */
-	  cpu = "skylake";
+	  cpu = astune ? "corei7" : "skylake";
 	  break;
 	case 0x55:
-	  if (has_avx512vnni)
+	  if (astune)
+	    cpu = "corei7";
+	  else if (has_avx512vnni)
 	    /* Cascade Lake.  */
 	    cpu = "cascadelake";
 	  else
@@ -841,15 +845,15 @@ const char *host_detect_local_cpu (int a
 	  break;
 	case 0x57:
 	  /* Knights Landing.  */
-	  cpu = "knl";
+	  cpu = astune ? "k1om" : "knl";
 	  break;
 	case 0x66:
 	  /* Cannon Lake.  */
-	  cpu = "cannonlake";
+	  cpu = astune ? "corei7" : "cannonlake";
 	  break;
 	case 0x85:
 	  /* Knights Mill.  */
-	  cpu = "knm";
+	  cpu =  astune ? "k1om" :"knm";
 	  break;
 	default:
 	  if (arch)
@@ -952,7 +956,7 @@ const char *host_detect_local_cpu (int a
 	    }
 	  else
 	    /* For -mtune, we default to -mtune=generic.  */
-	    cpu = "generic";
+	    cpu = astune ? NULL : "generic";
 	  break;
 	}
       break;
@@ -968,17 +972,17 @@ const char *host_detect_local_cpu (int a
 	cpu = "pentium4";
       break;
     case PROCESSOR_GEODE:
-      cpu = "geode";
+      cpu = astune ? NULL : "geode";
       break;
     case PROCESSOR_K6:
       if (arch && has_3dnow)
-	cpu = "k6-3";
+	cpu = astune ? "k6-2" : "k6-3";
       else
 	cpu = "k6";
       break;
     case PROCESSOR_ATHLON:
       if (arch && has_sse)
-	cpu = "athlon-4";
+	cpu = astune ? "athlon" : "athlon-4";
       else
 	cpu = "athlon";
       break;
@@ -998,7 +1002,7 @@ const char *host_detect_local_cpu (int a
 		cpu = "eden-x2";
 	      else
 		/* Default to k8 */
-		cpu = "k8";
+		cpu = "k8-sse3";
 	    }
 	  else if (has_sse3)
 	    cpu = "k8-sse3";
@@ -1065,7 +1069,7 @@ const char *host_detect_local_cpu (int a
 	    cpu = "pentium";
 	}
       else
-	cpu = "generic";
+	cpu = astune ? NULL : "generic";
     }
 
   if (arch)
@@ -1160,9 +1164,19 @@ const char *host_detect_local_cpu (int a
 			ptwrite,
 			NULL);
     }
+  else if (astune)
+    {
+/* Looks like gcc don't pass flag(s) to as after "native". Place it here. */
+      const char *sse2avx = has_avx ? " -msse2avx" : "";
+      
+      options = concat (options, sse2avx, NULL);
+    }
 
 done:
-  return concat (cache, "-m", argv[0], "=", cpu, options, NULL);
+  return cpu ? (astune ?
+	concat (cache, "-mtune=", cpu, options, NULL) :
+	concat (cache, "-m", argv[0], "=", cpu, options, NULL)) :
+		cache;
 }
 #else
 
--- a/gcc/config/i386/gnu-user64.h	2018-12-25 03:16:35.000000000 +0300
+++ b/gcc/config/i386/gnu-user64.h	2018-12-25 15:07:59.634645809 +0300
@@ -47,7 +47,7 @@ see the files COPYING3 and COPYING.RUNTI
 #endif
 
 #undef ASM_SPEC
-#define ASM_SPEC "%{" SPEC_32 ":--32} \
+#define ASM_SPEC ASM_SPEC_MTUNE "%{" SPEC_32 ":--32} \
  %{" SPEC_64 ":--64} \
  %{" SPEC_X32 ":--x32} \
  %{msse2avx:%{!mavx:-msse2avx}}"
--- a/gcc/config/i386/gnu-user.h	2018-12-25 03:16:35.000000000 +0300
+++ b/gcc/config/i386/gnu-user.h	2018-12-25 15:09:43.362640594 +0300
@@ -67,6 +67,7 @@ along with GCC; see the file COPYING3.
 
 #undef  ASM_SPEC
 #define ASM_SPEC \
+  ASM_SPEC_MTUNE \
   "--32 %{msse2avx:%{!mavx:-msse2avx}}"
 
 #undef  SUBTARGET_EXTRA_SPECS
--- a/gcc/config/i386/i386.h	2018-12-25 03:16:35.000000000 +0300
+++ b/gcc/config/i386/i386.h	2018-12-25 15:07:59.634645809 +0300
@@ -715,6 +715,13 @@ extern const char *host_detect_local_cpu
 %{mtune=native:%>mtune=native %:local_cpu_detect(tune)}"
 #endif
 #endif
+
+#if defined(HAVE_LOCAL_CPU_DETECT) && defined(HAVE_GNU_AS)
+#define ASM_SPEC_MTUNE "%{march=native:%{!mtune=*:%>march=native %:local_cpu_detect(astune)}} \
+ %{mtune=native:%>mtune=native %:local_cpu_detect(astune)}"
+#else
+#define ASM_SPEC_MTUNE ""
+#endif
 
 /* Target CPU builtins.  */
 #define TARGET_CPU_CPP_BUILTINS() ix86_target_macros ()
--- a/gcc/config/i386/mingw-w64.h	2018-12-25 03:16:35.000000000 +0300
+++ b/gcc/config/i386/mingw-w64.h	2018-12-25 15:07:59.634645809 +0300
@@ -40,7 +40,7 @@ along with GCC; see the file COPYING3.
 /* Enable multilib.  */
 
 #undef ASM_SPEC
-#define ASM_SPEC "%{m32:--32} %{m64:--64}"
+#define ASM_SPEC ASM_SPEC_MTUNE "%{m32:--32} %{m64:--64}"
 
 #undef LIB_SPEC
 #define LIB_SPEC "%{pg:-lgmon} %{" SPEC_PTHREAD1 ":-lpthread} " \
--- a/gcc/config/i386/x86-64.h	2018-12-25 03:16:35.000000000 +0300
+++ b/gcc/config/i386/x86-64.h	2018-12-25 15:07:59.635645809 +0300
@@ -49,7 +49,7 @@ see the files COPYING3 and COPYING.RUNTI
 #define WCHAR_TYPE_SIZE 32
 
 #undef ASM_SPEC
-#define ASM_SPEC "%{m16|m32:--32} %{m64:--64} %{mx32:--x32}"
+#define ASM_SPEC ASM_SPEC_MTUNE "%{m16|m32:--32} %{m64:--64} %{mx32:--x32}"
 
 #undef ASM_OUTPUT_ALIGNED_BSS
 #define ASM_OUTPUT_ALIGNED_BSS(FILE, DECL, NAME, SIZE, ALIGN) \
